infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/StreamsSnapshotWriter.java

```diff
@@ -0,0 +1,273 @@
+    @Getter
+    private PersistedWriterMetadata persistedWriterMetadata;
+    HashMap<UUID, UUID> uuidMap;
+    Phase phase;
+
+
+    // The sequence number of the message, it has received.
+    // It is expecting the message in order of the sequence.
+
+    public StreamsSnapshotWriter(CorfuRuntime rt, LogReplicationConfig config, PersistedWriterMetadata persistedWriterMetadata) {
+        this.rt = rt;
+        this.persistedWriterMetadata = persistedWriterMetadata;
+        streamViewMap = new HashMap<>();
+        uuidMap = new HashMap<>();
+        shadowMap = new HashMap<>();
+        phase = Phase.TransferPhase;
+
+        for (String stream : config.getStreamsToReplicate()) {
+            String shadowStream = stream + SHADOW_STREAM_NAME_SUFFIX;
+            UUID streamID = CorfuRuntime.getStreamID(stream);
+            UUID shadowID = CorfuRuntime.getStreamID(shadowStream);
+            uuidMap.put(streamID, shadowID);
+            uuidMap.put(shadowID, streamID);
+            streamViewMap.put(streamID, stream);
+            shadowMap.put(shadowID, shadowStream);
+        }
+    }
+
+    /**
+     * clear all tables registered
+     * TODO: replace with stream API
+     */
+    void clearTables() {
+        CorfuStoreMetadata.Timestamp timestamp = persistedWriterMetadata.getTimestamp();
+        long persistSiteConfigID = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID);
+        long persistSnapStart = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapStart);
+        long persitSeqNum = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapSeqNum);
+
+        //for transfer phase start
+        if (siteConfigID != persistSiteConfigID || srcGlobalSnapshot != persistSnapStart ||
+                (persitSeqNum + 1)!= recvSeq) {
+            log.warn("Skip current siteConfigID " + siteConfigID + " srcGlobalSnapshot " + srcGlobalSnapshot + " currentSeqNum " + recvSeq +
+                    " persistedMetadata " + persistedWriterMetadata.getSiteConfigID() + " startSnapshot " + persistedWriterMetadata.getLastSnapStartTimestamp() +
+                    " lastSnapSeqNum " + persistedWriterMetadata.getLastSnapSeqNum());
+            return;
+        }
+
+
+        TxBuilder txBuilder = persistedWriterMetadata.getTxBuilder();
+        persistedWriterMetadata.appendUpdate(txBuilder, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID, siteConfigID);
+
+
+        for (UUID streamID : streamViewMap.keySet()) {
+            UUID usedStreamID = streamID;
+            if (phase == Phase.TransferPhase) {
+                usedStreamID = uuidMap.get(streamID);
+            }
+
+            SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+            txBuilder.logUpdate(usedStreamID, entry);
+        }
+
+        txBuilder.commit(timestamp);
```

Comment:
processOpaqueEntry() wraps this in a try-catch block.  Why no exception handling here?